File: Paralelné Programovanie/Zadanie3/Uloha_3.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>

////////// PARAMETERS //////////

#define POCET_ZAUJEMCOV 25
#define POCET_ZA_CAS 2
#define CAS_ZAUJEMCA 2

#define POCET_LUKOV 3
#define MAX_POCET_STRELNICA 5
int AKTUALNY_POCET_LUKOV = POCET_LUKOV;
int AKTUALNY_POCET_STRELNICA = 0;

int POZICIE[MAX_POCET_STRELNICA] = {0};
int POCET_NA_POZICII[MAX_POCET_STRELNICA] = {0};
// kladne cisla - niekto striela, zaporne cisla - niekto zbiera
int STATE = 0;

int LUK_COUNT = 0;
int KUSA_COUNT = 0;

#define KUSA_CAS 5
#define LUK_CAS 2

// 1 - luk, 0 - kusa
int zbran_cas[2] = {KUSA_CAS, LUK_CAS};

pthread_mutex_t vstup_mutex = PTHREAD_MUTEX_INITIALIZER;
pthread_cond_t vstup_cond = PTHREAD_COND_INITIALIZER;
pthread_mutex_t zbrane_mutex = PTHREAD_MUTEX_INITIALIZER;
pthread_mutex_t strielanie_zbieranie_mutex = PTHREAD_MUTEX_INITIALIZER;
pthread_cond_t strielanie_zbieranie_cond = PTHREAD_COND_INITIALIZER;


int zober_zbran() {

    pthread_mutex_lock(&zbrane_mutex);

    int zbran;

    if (AKTUALNY_POCET_LUKOV > 0) {
        --AKTUALNY_POCET_LUKOV;
        ++LUK_COUNT;
        zbran = 1;
    }
    
    else {
        ++KUSA_COUNT;
        zbran = 0;
    }

    pthread_mutex_unlock(&zbrane_mutex);
    return zbran;
}


// return pozicia na strelnici
int vstup_na_strelnicu() {

    pthread_mutex_lock(&vstup_mutex);

    while (AKTUALNY_POCET_STRELNICA == MAX_POCET_STRELNICA) {
        pthread_cond_wait(&vstup_cond, &vstup_mutex);
    }

    int index;

    for (int i = 0; i < MAX_POCET_STRELNICA; i++) {
        if (!POZICIE[i]) {
            index = i;
            break;
        }
    }

    POZICIE[index] = 1;
    ++POCET_NA_POZICII[index];
    ++AKTUALNY_POCET_STRELNICA;

    pthread_mutex_unlock(&vstup_mutex);

    return index;
}


void strielaj(int zbran) {

    pthread_mutex_lock(&strielanie_zbieranie_mutex);

    // niekto zbiera, treba pockat
    while (STATE < 0) {
        pthread_cond_wait(&strielanie_zbieranie_cond, &strielanie_zbieranie_mutex);
    }

    ++STATE;
    pthread_cond_signal(&strielanie_zbieranie_cond);
    pthread_mutex_unlock(&strielanie_zbieranie_mutex);

    printf("SHOOTING\n");
    sleep(zbran_cas[zbran]);
    printf("END SHOOTING\n");

    // prestan strielat
    pthread_mutex_lock(&strielanie_zbieranie_mutex);
    --STATE;
    pthread_cond_signal(&strielanie_zbieranie_cond);
    pthread_mutex_unlock(&strielanie_zbieranie_mutex);
}


void zbieraj(int cas) {

    pthread_mutex_lock(&strielanie_zbieranie_mutex);

    // niekto striela, treba pockat
    while (STATE > 0) {
        pthread_cond_wait(&strielanie_zbieranie_cond, &strielanie_zbieranie_mutex);
    }

    --STATE;
    pthread_cond_signal(&strielanie_zbieranie_cond);
    pthread_mutex_unlock(&strielanie_zbieranie_mutex);

    printf("COLLECTING\n");
    sleep(cas);
    printf("END COLLECTING\n");

    // dozbieraj
    pthread_mutex_lock(&strielanie_zbieranie_mutex);
    ++STATE;
    pthread_cond_signal(&strielanie_zbieranie_cond);
    pthread_mutex_unlock(&strielanie_zbieranie_mutex);
}    


void odid(int pozicia) {

    pthread_mutex_lock(&vstup_mutex);

    POZICIE[pozicia] = 0;
    --AKTUALNY_POCET_STRELNICA;

    pthread_cond_signal(&vstup_cond);
    pthread_mutex_unlock(&vstup_mutex);
}


void vrat_zbran(int zbran) {

    pthread_mutex_lock(&zbrane_mutex);

    if (zbran) {
        ++AKTUALNY_POCET_LUKOV;
    }

    pthread_mutex_unlock(&zbrane_mutex);
}


void zaujemca() {
    
    int zbran = zober_zbran();
    int pozicia = vstup_na_strelnicu();
    strielaj(zbran);
    zbieraj(pozicia + 1);
    odid(pozicia);
    vrat_zbran(zbran);
}


int main() {

    pthread_t zaujemci[POCET_ZAUJEMCOV];

    for (int i = 0; i < POCET_ZAUJEMCOV; i++) {
        if (i && i % POCET_ZA_CAS == 0) {
            sleep(CAS_ZAUJEMCA);
        }

        pthread_create(&zaujemci[i], NULL, &zaujemca, NULL);
    }

    for (int i = 0; i < POCET_ZAUJEMCOV; i++) {
        pthread_join(zaujemci[i], NULL);
    }

    printf("Z luku strielalo: %d zaujemcov\n", LUK_COUNT);
    printf("Z kuse strielalo: %d zaujemcov\n", KUSA_COUNT);

    printf("Pocet strelcov z pozicie:\n");
    
    for (int i = 0; i < MAX_POCET_STRELNICA; i++) {
        printf("miesto %d: strelcov: %d\n", i + 1, POCET_NA_POZICII[i]);
    }

    return 0;
}

File: Paralelné Programovanie/Zadanie3/Uloha_3.py
from time import sleep, time
from threading import Thread, Lock, Condition


########## PARAMETERS ##########
POCET_ZAUJEMCOV = 25
POCET_ZA_CAS = 2
CAS_ZAUJEMCA = 4

POCET_LUKOV = 3
MAX_POCET_STRELNICA = 5

POZICIE = [False] * MAX_POCET_STRELNICA
POCET_NA_POZICII = [0] * MAX_POCET_STRELNICA
# kladne cisla - niekto striela, zaporne cisla - niekto zbiera
STATE = 0

LUK_COUNT = 0
KUSA_COUNT = 0

KUSA_CAS = 5
LUK_CAS = 2

# 1 - luk, 0 - kusa
zbran_cas = {
    0: KUSA_CAS,
    1: LUK_CAS
}

# ak Condition nedostane ako parameter mutex, vytvori si vlastny
vstup_cond = Condition()
zbrane_mutex = Lock()
strielanie_zbieranie_cond = Condition()


def zober_zbran():

    global zbrane_mutex, POCET_LUKOV, LUK_COUNT, KUSA_COUNT

    zbrane_mutex.acquire()

    if POCET_LUKOV > 0:
        POCET_LUKOV -=1
        LUK_COUNT += 1
        zbran = 1

    else:
        KUSA_COUNT += 1
        zbran = 0

    zbrane_mutex.release()
    return zbran


# return pozicia na strelnici
def vstup_na_strelnicu():

    global vstup_cond, POZICIE

    vstup_cond.acquire()

    while all(POZICIE):
        vstup_cond.wait()
    
    index = POZICIE.index(False)
    POZICIE[index] = True
    POCET_NA_POZICII[index] += 1
    vstup_cond.release()

    return index


def strielaj(zbran):

    global strielanie_zbieranie_cond, STATE

    strielanie_zbieranie_cond.acquire()

    # niekto zbiera, treba pockat
    while STATE < 0:
        strielanie_zbieranie_cond.wait()
    
    STATE += 1
    strielanie_zbieranie_cond.notify()
    strielanie_zbieranie_cond.release()

    print("SHOOTING")
    sleep(zbran_cas[zbran])
    print("END SHOOTING")

    # prestan strielat
    strielanie_zbieranie_cond.acquire()
    STATE -= 1
    strielanie_zbieranie_cond.notify()
    strielanie_zbieranie_cond.release()


def zbieraj(cas):

    global strielanie_zbieranie_cond, STATE, POZICIE

    strielanie_zbieranie_cond.acquire()

    # niekto striela, treba pockat
    while STATE > 0:
        strielanie_zbieranie_cond.wait()
    
    STATE -= 1
    strielanie_zbieranie_cond.notify()
    strielanie_zbieranie_cond.release()

    print("COLLECTING")
    sleep(cas)
    print("END COLLECTING")

    # dozbieraj
    strielanie_zbieranie_cond.acquire()
    STATE += 1
    strielanie_zbieranie_cond.notify()
    strielanie_zbieranie_cond.release()


def odid(pozicia):

    global vstup_cond, POZICIE

    vstup_cond.acquire()

    POZICIE[pozicia] = False
    vstup_cond.notify()
    vstup_cond.release()


def vrat_zbran(zbran):

    global zbrane_mutex, POCET_LUKOV
    
    zbrane_mutex.acquire()
    
    if zbran:
        POCET_LUKOV += 1
    
    zbrane_mutex.release()


def zaujemca():

    zbran = zober_zbran()
    pozicia = vstup_na_strelnicu()
    strielaj(zbran)
    zbieraj(pozicia + 1)
    odid(pozicia)
    vrat_zbran(zbran)


def main():

    zaujemci = []

    start_time = time()

    for i in range(POCET_ZAUJEMCOV):
        if i and i % POCET_ZA_CAS == 0:
            sleep(CAS_ZAUJEMCA)

        zaujemci.append(Thread(target=zaujemca))
        zaujemci[i].start()
    
    for i in range(POCET_ZAUJEMCOV):
        zaujemci[i].join()
    
    print("Všetci záujemci sa vystriedali za: {:.2f} sekúnd".format(round(time() - start_time, 2)))
    print("Z luku strieľalo: %d záujemcov" % (LUK_COUNT))
    print("Z kuše strieľalo: %d záujemcov" % (KUSA_COUNT))

    print("Pocet strelcov z pozicie:")
    
    for i, v in enumerate(POCET_NA_POZICII):
        print("miesto %d: strelcov: %d" % (i + 1, v))


main()
